### ISSUE TITLE
Support `Cflags.private`

### ISSUE THREAD

**Issue by shithappens2016** [mesonbuild/meson#10881]
Meson's pkgconfig module does not seem to support `Cflags.private`. This feature is becoming more and more useful.Please consider supporting it.

**Comment by eli-schwartz** [comment]
This is item number 3 at https://github.com/mesonbuild/meson/issues/3111

The problem is the official pkg-config spec doesn't support Cflags.private but OTOH freedesktop pkg-config hasn't added it basically because, well, freedesktop pkg-config is basically dead.

It's noted in the `pc(5)` manpage as a pkgconf extension.


IIRC the plan for pkgconf is to specify its behavior and declare it the canonical version on the grounds that it has taken over mindshare and is the only maintained implementation, @kaniini would know more.

**Comment by xclaesse** [comment]
Given that meson cannot build both shared and static lib for Windows, cflags.private is pretty useless. If you build the static lib only, you can put it in cflags.

We would have to check how legacy pkg-config deals if unknowns field. If it ignore them, we can just add those extra field anyway.

Another way is to merge my implementation of pkg-config and just ignore legacy pkg-config completely.

**Comment by xclaesse** [comment]
Btw one thing I've been wanting for years is required.internal, it's part of pkgconf IIRC.

**Comment by guijan** [comment]
This issue should be reopened, `Cflags.private` is important for writing Windows libraries. In fact, that seems to be the reason it was added to pkgconf: https://github.com/pkgconf/pkgconf/pull/50

**Comment by kasper93** [comment]
> Given that meson cannot build both shared and static lib for Windows, cflags.private is pretty useless. If you build the static lib only, you can put it in cflags.

I'm just looking and my `default_library=both` and scratching my head how to produce proper pc file for this build. Both gcc and clang can build both static and shared lib on windows. Not sure about MSVC, but clang is good enough reason to support this feature.

**Comment by guijan** [comment]
You will have issues with that because either the shared or the static library won't export symbols right. You can write your own Meson code with both a `static_library()` and a `shared_library()` call and toggle them depending on what `default_library` is set to which will compile the program twice in the case of `default_library == 'both'` and allows you to set the right macro to make your code export Windows libraries as it should, but then you'll be unable to make the corresponding .pc file because there's no support for `Cflags.private` in Meson.

**Comment by eli-schwartz** [comment]
> You will have issues with that because either the shared or the static library won't export symbols right. You can write your own Meson code with both a `static_library()` and a `shared_library()` call 

My understanding is that vs_module_defs makes this work; it's an alternative to dllexport that is applied by the linker when creating a DLL but isn't used when archiving a static library.

> but then you'll be unable to make the corresponding .pc file because there's no support for `Cflags.private` in Meson.

There's actually a pretty good summary of this at https://github.com/libass/libass/commit/c5ee66dfb4b10c3c499aa804c6d1abb195c696e1 but tl;dr you do not actually need dllimport  when using a shared library, at least for functions. It just makes it a bit faster. So you don't need any pkg-config CFLAGS in that case.

**Comment by kasper93** [comment]
> You will have issues with that because either the shared or the static library won't export symbols right.

Citation needed. Decorating function with dllexport adds them to export table, but the function itself doesn't change. When building static lib it may be not desirable that they got exported when linking to other shared library, but this is what linux does also when you not use `--exclude-libs`.

Only problems you can have when linking to shared lib and WANT to export something from static one, but if the object is not referenced it wouldn't be exported because linker would not look at export table, but it is completely different problem and works just fine when using object files directly.

For building with different flags this would be needed https://github.com/mesonbuild/meson/issues/3304 but ultimately is is not required.

> tl;dr you do not actually need dllimport when using a shared library, at least for functions. It just makes it a bit faster. So you don't need any pkg-config CFLAGS in that case.

Yes, `dllimport` is an optimization. But whether someone want to use it or not is not a reason to reject this feature request.

As a fun note this works in current meson, just forcefully injecting `Cflags.private:`  `extra_cflags: ['\nCflags.private:', '-DFOO_STATIC']`

**Comment by eli-schwartz** [comment]
> Yes, `dllimport` is an optimization. But whether someone want to use it or not is not a reason to reject this feature request.

It's also only a sometimes-optimization, since you can only do it for function symbols, not for object symbols -- in the latter case, dllimport is mandatory and you still need Cflags.private.

Either way, I agree regardless that this feature request has merit, even as an optimization.

**Comment by redcode** [comment]
I'm using `Cflags.private` in my projects (example: [Z80.pc.in](https://github.com/redcode/Z80/blob/master/support/Z80.pc.in) in [redcode/Z80](https://github.com/redcode/Z80)]. It would be very handy for building static libraries on Windows.
